shark_turbine/kernel/wave/codegen.py

```diff
@@ -200,76 +201,135 @@ def add_emitter_subs(emitter: WaveEmitter) -> dict[IndexSymbol, Any]:
     return dynamics
 
 
+_Rational = namedtuple("_Rational", ["numerator", "denominator"])
+
+
 def gen_sympy_index(dynamics: dict[IndexSymbol, Any], expr: sympy.Expr) -> OpResult:
     stack: list[OpResult] = []
 
-    def _broadcast(a, b):
-        if not isinstance(a, (Value, OpResult)):
-            a = a.result
+    def _get_ir_value(arg):
+        if not isinstance(arg, (Value, OpResult)):
+            arg = arg.result
+
+        return arg
 
-        if not isinstance(b, (Value, OpResult)):
-            b = b.result
+    def _check_vec_scalar(a, b):
+        return isinstance(a.type, VectorType) and a.type.element_type == b.type
+
+    def _broadcast(a, b):
+        a = _get_ir_value(a)
+        b = _get_ir_value(b)
 
         if a.type == b.type:
             return a, b
 
-        if isinstance(a.type, VectorType) and isinstance(
-            b.type, (IndexType, IntegerType)
-        ):
-            assert a.type.element_type == b.type
+        if _check_vec_scalar(a, b):
             b = vector_d.splat(a.type, b)
             return a, b
 
-        if isinstance(a.type, (IndexType, IntegerType)) and isinstance(
-            b.type, VectorType
-        ):
-            assert b.type.element_type == a.type
+        if _check_vec_scalar(b, a):
             a = vector_d.splat(b.type, a)
             return a, b
 
         raise CodegenError(f"Cannot broadcast {a.type} and {b.type}")
 
-    def _process_mul_add_ops(term, is_mul):
-        args = []
-        callables = []
-        for _ in range(len(term.args)):
-            val = stack.pop()
-            if callable(val):
-                callables.append(val)
-            else:
-                args.append(val)
-        operation = None
-        for arg in args:
-            if operation is None:
-                operation = arg
-                continue
+    def get_const_val(arg):
+        if isinstance(arg, OpResult):
+            arg = arg.owner.opview
 
-            if is_mul:
-                operation = arith_d.MulIOp(*_broadcast(operation, arg))
-            else:
-                operation = arith_d.AddIOp(*_broadcast(operation, arg))
+        if isinstance(arg, arith_d.ConstantOp):
+            value = arg.attributes["value"]
+            if isinstance(value, IntegerAttr):
+                return int(value)
```

Comment:
Throw error if it is not an int?

Reply:
We can actually encounter vector constants now, so this code is intended to fall through to the `return None`